[Navigation] Track in-progress navigations

TingPing opened the pull request:
#### a09b94a65009978eb3e5f4417b46f8687c834564
<pre>
[Navigation] Track in-progress navigations
<a href="https://bugs.webkit.org/show_bug.cgi?id=272459">https://bugs.webkit.org/show_bug.cgi?id=272459</a>

Reviewed by Alex Christensen.

This makes use of the NavigationAPITracker to correctly complete
the Navigation::Result when events happen.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/commit-behavior/after-transition-uncancelable-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-noop-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-navigation-back-same-document-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/same-url-replace-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/same-url-replace-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-base-url-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-gc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-src-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-replace-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-already-detached-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-detach-in-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-already-detached-expected.txt:

* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::callPromisePairFunction):

The JSPromise needs to not be cleared because we
create a DOMPromise with it later.

* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::createDOMPromise):
(WebCore::Navigation::createErrorResult):
(WebCore::Navigation::maybeSetUpcomingNonTraversalTracker):
(WebCore::Navigation::addUpcomingTrarveseAPIMethodTracker):
(WebCore::Navigation::apiMethodTrackerDerivedResult):
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::updateCurrentEntry):
(WebCore::Navigation::hasEntriesAndEventsDisabled const):
(WebCore::Navigation::resolveFinishedPromise):
(WebCore::notifyCommittedToEntry):
(WebCore::Navigation::updateForNavigation):
(WebCore::Navigation::promoteUpcomingAPIMethodTracker):
(WebCore::Navigation::cleanupAPIMethodTracker):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::createErrorResult): Deleted.
* Source/WebCore/page/Navigation.h:
(WebCore::NavigationAPIMethodTracker::NavigationAPIMethodTracker):

Canonical link: <a href="https://commits.webkit.org/277688@main">https://commits.webkit.org/277688@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e69879d1f26ecfdac04b24bbbd53061ed06c3e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51235 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50969 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44346 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50586 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25014 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39443 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48863 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25188 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20589 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/22668 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6337 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44626 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43347 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52873 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23328 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19680 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46781 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24593 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41895 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45695 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10653 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25398 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24316 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -121,14 +124,21 @@ enum EventTargetInterfaceType Navigation::eventTargetInterface() const
     return EventTargetInterfaceType::Navigation;
 }
 
+RefPtr<DOMPromise> Navigation::createDOMPromise(Ref<DeferredPromise>& deferredPromise)
```
This likely does not need to be a member function.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -171,34 +180,65 @@ NavigationAPIMethodTracker Navigation::maybeSetUpcomingNonTraversalTracker(Ref<D
-    auto newURL = URL { currentURL, url };
+
+    if (!frame())
+        return createErrorResult(WTFMove(committed), WTFMove(finished), ExceptionCode::InvalidStateError, "Invalid state"_s);
```
It is probably if I move these kind of detach fixes to https://github.com/WebKit/WebKit/pull/25013. WDYT?

TingPing replied:
Yeah that would be fine.

rwlbuis commented on Source/WebCore/page/Navigation.h:
```diff
@@ -46,11 +45,13 @@ enum class FrameLoadType : uint8_t;
-        , committedPromise(DOMPromise::create(*committed->globalObject(), *JSC::jsCast<JSC::JSPromise*>(committed->promise())))
-        , finishedPromise(DOMPromise::create(*finished->globalObject(), *JSC::jsCast<JSC::JSPromise*>(finished->promise())))
+        , committedPromise(&committed.leakRef())
+        , finishedPromise(&finished.leakRef())
```
Should this just use WTFMove?

TingPing replied:
It is transferring ownership of a `Ref` to a `RefPtr`. I saw this pattern in a few places.

achristensen07 replied:
Yeah, this looks like a memory leak as it stands now.

rwlbuis replied:
Great, the important thing is indeed that NavigationAPIMethodTracker takes ownership.

cdumez replied:
God please don't do this lol. use `WTFMove()`. I'm very curious where you saw this "pattern".

TingPing replied:
I don't remember, it's used ~121 times in the codebase.

This is a move constructor for RefPtr:

```cpp
RefPtr<T, U, V>::RefPtr(Ref<X, Y>&& reference)
    : m_ptr(&reference.leakRef())
```

I understand that this wouldn't ref while my usage would though. My bad.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -171,37 +184,61 @@ NavigationAPIMethodTracker Navigation::maybeSetUpcomingNonTraversalTracker(Ref<D
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-navigate
-Navigation::Result Navigation::navigate(const String& url, NavigateOptions&& options, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
+Navigation::Result Navigation::navigate(const String& url, NavigateOptions&& options, Ref<DeferredPromise>&&committed, Ref<DeferredPromise>&&finished)
```
I guess this change was not intentional?

rwlbuis commented on Source/WebCore/page/Navigation.h:
```diff
@@ -144,13 +145,21 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
     void derefEventTarget() final { deref(); }
 
+    Result createErrorResult(Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished, Exception&&);
+    Result createErrorResult(Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished, ExceptionCode, const String& errorMessage);
```
Why do these need to be methods and not static functions?

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1290,6 +1290,10 @@ void FrameLoader::loadInSameDocument(URL url, RefPtr<SerializedScriptValue> stat
+
+        // During the NavigateEvent the page could have been detached from JS.
+        if (!m_frame->page())
+            return;
```
Is this check needed now or also in the past? Any test that used tobe affected?

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -51,6 +52,8 @@
 
 namespace WebCore {
 
+static uint64_t lastTrackerID;
```
This should be initialized, probably to 0.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -171,37 +184,61 @@ NavigationAPIMethodTracker Navigation::maybeSetUpcomingNonTraversalTracker(Ref<D
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-method-tracker-derived-result
+Navigation::Result Navigation::apiMethodTrackerDerivedResult(NavigationAPIMethodTracker& apiMethodTracker)
```
Is apiMethodTracker actually being changed?

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -396,6 +456,35 @@ static bool documentCanHaveURLRewritten(const RefPtr<Document>& document, const
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-method-tracker-clean-up
+void Navigation::cleanupAPIMethodTracker(NavigationAPIMethodTracker& apiMethodTracker)
```
Is apiMethodTracker actually being changed?

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -121,14 +124,21 @@ enum EventTargetInterfaceType Navigation::eventTargetInterface() const
     return EventTargetInterfaceType::Navigation;
 }
 
+static RefPtr<DOMPromise> createDOMPromise(Ref<DeferredPromise>& deferredPromise)
```
This can likely be const &.

cdumez replied:
Neither seems right. We only pass smart pointers to transfer ownership. So either this remains a smart pointer and we should use a r-value reference `Ref&&`, or this should be using a `DeferredPromise&`.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -343,6 +383,25 @@ bool Navigation::hasEntriesAndEventsDisabled() const
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#notify-about-the-committed-to-entry
+static void notifyCommittedToEntry(NavigationAPIMethodTracker& apiMethodTracker, RefPtr<NavigationHistoryEntry> entry)
```
The entry param likely violates the rule Chris mentioned: "We only pass smart pointers to transfer ownership".

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -211,51 +248,62 @@ Navigation::Result Navigation::navigate(const String& url, NavigateOptions&& opt
         m_upcomingNonTraverseMethodTracker = std::nullopt;
+        // Ref commitedPromise = apiMethodTracker.committedPromise.releaseNonNull();
+        // Ref finishedPromise = apiMethodTracker.finishedPromise.releaseNonNull();
+        // return createErrorResult(WTFMove(commitedPromise), WTFMove(finishedPromise), ExceptionCode::AbortError, "Navigation wasn't completed"_s);
```
Is this safe to enable? If so I'd enable it, otherwise I'd remove it since the rule is to not include commented code.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -211,51 +248,59 @@ Navigation::Result Navigation::navigate(const String& url, NavigateOptions&& opt
-    frame()->loader().load(FrameLoadRequest(*frame(), URL(entry.url())));
+    auto entry = findEntryByKey(key);
+    ASSERT(entry);
+    frame()->loader().load(FrameLoadRequest(*window()->frame(), URL(entry.value()->url())));
```
IIRC we can leave off window() from window()->frame().

rwlbuis reviewed (commented):
Patch looks in a good shape to me now.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -396,6 +454,35 @@ static bool documentCanHaveURLRewritten(const RefPtr<Document>& document, const
+    else {
+        auto& key = apiMethodTracker.key;
+        // FIXME: ASSERT(!key.isNull());
+        // FIXME: ASSERT(m_upcomingTraverseMethodTrackers.contains(key));
```
We generally don't commit commented code in WebKit, but a text description of things we still need to do is fine.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -396,6 +454,35 @@ static bool documentCanHaveURLRewritten(const RefPtr<Document>& document, const
+    ASSERT(!m_ongoingAPIMethodTracker);
+
+    if (!destinationKey.isNull()) {
+        // FIXME: ASSERT(!m_upcomingNonTraverseMethodTracker);
```
// FIXME: We ought to be able to assert that m_upcomingNonTraverseMethodTracker is nullopt here.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -211,51 +248,59 @@ Navigation::Result Navigation::navigate(const String& url, NavigateOptions&& opt
     // FIXME: This is not a proper Navigation API initiated traversal, just a simple load for now.
     frame()->loader().load(FrameLoadRequest(*frame(), newURL));
 
+    // If the load never made it to the navigate event.
```
This comment isn't very clear.

TingPing committed:
[Navigation] Track in-progress navigations
https://bugs.webkit.org/show_bug.cgi?id=272459

Reviewed by Alex Christensen.

This makes use of the NavigationAPITracker to correctly complete
the Navigation::Result when events happen.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/commit-behavior/after-transition-uncancelable-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-crossorigin-sameorigindomain.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-window/click-samedocument-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-noop-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-navigation-back-same-document-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/same-url-replace-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/same-url-replace-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-base-url-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-gc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-src-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-replace-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-already-detached-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-detach-in-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-already-detached-expected.txt:

* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::callPromisePairFunction):

The JSPromise needs to not be cleared because we
create a DOMPromise with it later.

* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::createDOMPromise):
(WebCore::Navigation::createErrorResult):
(WebCore::Navigation::maybeSetUpcomingNonTraversalTracker):
(WebCore::Navigation::addUpcomingTrarveseAPIMethodTracker):
(WebCore::Navigation::apiMethodTrackerDerivedResult):
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::updateCurrentEntry):
(WebCore::Navigation::hasEntriesAndEventsDisabled const):
(WebCore::Navigation::resolveFinishedPromise):
(WebCore::notifyCommittedToEntry):
(WebCore::Navigation::updateForNavigation):
(WebCore::Navigation::promoteUpcomingAPIMethodTracker):
(WebCore::Navigation::cleanupAPIMethodTracker):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::createErrorResult): Deleted.
* Source/WebCore/page/Navigation.h:
(WebCore::NavigationAPIMethodTracker::NavigationAPIMethodTracker):

Canonical link: https://commits.webkit.org/277688@main
